references/settings.md
# Settings

To create a Settings object, you need to always specify name (the name of the project as a unique identifier) and folder (the folder location to save/read the settings file), or set the environment variables DEDUPER_NAME and DEDUPER_FOLDER. 

If a settings file already exists, you can read in all other settings with the read function. If not, you can create and save your settings to file with the save function. 

You can specify some or all settings when you create a Settings object, or set settings later with the set function. With the sync function you can either save settings to file if it doesn't exist or if you specified other settings, or read settings from file if it does exist.

#### Example of creating settings:

```
attributes = ["givenname", "surname", "suburb", "postcode"]


settings = Settings(
        name="default",  # the name of the project, a unique identifier
        folder=".././.dedupe",  # path to folder where settings and data will be saved
        attributes=["givenname", "surname", "suburb", "postcode"],  # list of entity attribute names
        model=SettingsModel(
            dedupe=True,
            n=1000,
            k=3,
            max_compare=20_000,
            n_covered=5_000,
            cpus=20,  # parallelize distance computations
            path_model="./.dedupe/test_model",  # where to save the model
        ),
        db=SettingsDB(
            path_database="postgresql+psycopg2://username:password@172.22.39.26:8000/db",  # where to save the sqlite database holding intermediate data
            db_schema="dedupe",
        ),
        label_studio=SettingsLabelStudio(
            api_key="33344e8a477f8adc3eb6aa1e41444bde76285d96",  # label studio port
            description="chansoo test project",  # label studio description of project
            port=8089,  # label studio port
        ),
        fast_api = SettingsService(
            port=8090,  # fast api port
        )
    )
```

#### If saved:

a settings file is saved to deduper_settings_{name}.json

```
{
  "name": "test",
  "folder": ".dedupe",
  "other": {
    "attributes": [
      "givenname",
      "surname",
      "suburb",
      "postcode"
    ],
    "cpus": 15,
    "path_database": ".dedupe/test.db",
    "path_model": ".dedupe/test_model",
    "label_studio": {
      "host": "http://0.0.0.0",
      "port": 8089,
      "api_key": "bc66ff77abeefc91a5fecd031fc0c238f9ad4814",
      "description": "gs test project"
    },
    "fast_api": {
      "host": "http://0.0.0.0",
      "port": 8003
    }
  }
}
```

references/configuration.md
# Configuration

Deduper requires a config file, which is loaded when deduper is initialized, e.g. `import dedupe`.

Currently, a distinct configuration file is required for each dedupe application (issue created: https://github.com/chansooligans/deduper/issues/28)

There are two options:

(1) A global config file at ~/.dedupe/config.ini
(2) A per-project config file at $CWD/.dedupe/config.ini, where $CWD is the folder you're running your script from (the folder where you `import dedupe`).

The config file must be called `config.ini` contain:

```
[MAIN]
HOST = http://172.22.39.26
MODEL_FILEPATH = [path to model to load, or save if it does not exist]
CACHE_FILEPATH = [path to store database files for SQLite and comparison pairs]

[LABEL_STUDIO]
PORT = 8001
API_KEY = [API KEY from Label Studio interface after signing up / logging in]
TITLE = [Title of Dedupe Project]
DESCRIPTION = [Description of Project]

[FAST_API]
PORT = 8000
CPUS = [Number of CPUs for distance calcs and ML models]
```

dedupe/config.py
import configparser
import os
import logging

CONFIG_FOLDER_NAME = ".dedupe"

def get_dedupe_file_path(*filepath) -> str:
    home = os.path.expanduser("~")
    if home is None:
        raise RuntimeError("No home directory.")

    return os.path.join(home, CONFIG_FOLDER_NAME, *filepath)

def get_project_dedupe_file_path(*filepath):
    """Return the full path to a filepath in ${CWD}/.dedupe.
    """
    return os.path.join(os.getcwd(), CONFIG_FOLDER_NAME, *filepath)

CONFIG_FILENAMES = [
    get_dedupe_file_path("config.ini"),
    get_project_dedupe_file_path("config.ini"),
]

class Config:

    def __init__(self):
        
        self.config = configparser.ConfigParser()
        
        for filename in CONFIG_FILENAMES:
            if os.path.exists(filename):
                self.config.read(filename)
                break
        
        if self.config.sections() != ['MAIN', 'LABEL_STUDIO', 'FAST_API']:
            logging.error(
                """
                A config file is required, there are two options:

                (1) A global config file at ~/.dedupe/config.ini
                (2) A per-project config file at $CWD/.dedupe/config.ini, where $CWD is the folder you're running your script from (the folder where you `import dedupe`).

                The config file should contain:

                [MAIN]
                HOST = http://172.22.39.26
                MODEL_FILEPATH = [path to model to load, or save if it does not exist]
                CACHE_FILEPATH = [path to store database files for SQLite and comparison pairs]

                [LABEL_STUDIO]
                PORT = 8001
                API_KEY = [API KEY from Label Studio interface after signing up / logging in]
                TITLE = [Title of Dedupe Project]
                DESCRIPTION = [Description of Project]

                [FAST_API]
                PORT = 8000
                CPUS = [Number of CPUs for distance calcs and ML models]
                """
            )

    @property
    def host(self):
        return self.config["MAIN"]["HOST"]

    @property
    def ls_url(self):
        return f"""{self.host}:{self.config["LABEL_STUDIO"]["PORT"]}"""

    @property
    def ls_api_key(self):
        return self.config["LABEL_STUDIO"]["API_KEY"]

    @property
    def ls_title(self):
        return self.config["LABEL_STUDIO"]["TITLE"]

    @property
    def ls_description(self):
        return self.config["LABEL_STUDIO"]["DESCRIPTION"]

    @property
    def fast_api_port(self):
        return self.config["FAST_API"]["PORT"]
    
    @property
    def fast_api_url(self):
        return f"""{self.host}:{self.config["FAST_API"]["PORT"]}"""

    @property
    def cpus(self):
        return int(self.config["FAST_API"]["CPUS"])

    @property
    def model_fp(self):
        return self.config["MAIN"]["MODEL_FILEPATH"]

    @property
    def cache_fp(self):
        return self.config["MAIN"]["CACHE_FILEPATH"]



